crates/bindings-csharp/BSATN.Codegen/Diag.cs

```diff
@@ -0,0 +1,162 @@
+namespace SpacetimeDB.Codegen;
+
+using System.Linq.Expressions;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using static SpacetimeDB.Codegen.Utils;
+
+public record ErrorDescriptorGroup(string Tag, string Name)
+{
+    private int idCounter = 0;
+
+    public string NextId() => $"{Tag}{++idCounter:D4}";
+}
+
+// Roslyn `Diagnostics` requires declaration using old-school string format ("foo {0} bar {1}")
+// which is not very type-safe when you want to pass arguments from arbitrary places in the codebase.
+//
+// This helper class allows us to define diagnostics in a type-safe way by using LINQ expressions
+// to extract the format and arguments from a normal-looking lambda.
```

Comment:
Could you add examples of the two styles of usage?

Reply:
Do you mean the regular `Diagnostic` constructor vs this helper? Or which 2 styles?